app/src/main/java/io/appium/uiautomator2/model/UiElementSnapshot.java

```diff
@@ -38,125 +42,183 @@
-        if (selectionRange != null) {
-            attributes.put(Attribute.SELECTION_START, selectionRange.getLower());
-            attributes.put(Attribute.SELECTION_END, selectionRange.getUpper());
-        }
-        setAttribute(attributes, Attribute.SELECTED, node.isSelected());
-        setAttribute(attributes, Attribute.BOUNDS, AxNodeInfoHelper.getBounds(node).toShortString());
-        setAttribute(attributes, Attribute.DISPLAYED, node.isVisibleToUser());
-        // Skip CONTENT_SIZE as it is quite expensive to compute it for each element
-        this.attributes = Collections.unmodifiableMap(attributes);
+        if (includedAttributes != null) {
+            // Class name attribute should always be there
+            this.includedAttributes.add(Attribute.CLASS);
+            this.includedAttributes.addAll(includedAttributes);
+        }
+        this.attributes = collectAttributes(node, index);
         this.children = buildChildren(node);
     }
 
-    private UiElementSnapshot(AccessibilityNodeInfo node, int index) {
-        this(node, index, DEFAULT_MAX_DEPTH);
+    private UiElementSnapshot(AccessibilityNodeInfo node, int index, @Nullable Set<Attribute> includedAttributes) {
+        this(node, index, DEFAULT_MAX_DEPTH, includedAttributes);
     }
 
-    private UiElementSnapshot(String hierarchyClassName, AccessibilityNodeInfo[] childNodes, int index) {
+    private UiElementSnapshot(String hierarchyClassName, AccessibilityNodeInfo[] childNodes, int index,
+                              @Nullable Set<Attribute> includedAttributes) {
         super(null);
         Map<Attribute, Object> attribs = new LinkedHashMap<>();
         setAttribute(attribs, Attribute.INDEX, index);
         setAttribute(attribs, Attribute.CLASS, hierarchyClassName);
         this.attributes = Collections.unmodifiableMap(attribs);
-        List<UiElementSnapshot> children = new ArrayList<>();
-        for (AccessibilityNodeInfo childNode : childNodes) {
-            children.add(new UiElementSnapshot(childNode, children.size()));
+        List<UiElementSnapshot> children = new ArrayList<>(childNodes.length);
+        for (int childNodeIdx = 0; childNodeIdx < childNodes.length; ++childNodeIdx) {
+            children.add(new UiElementSnapshot(childNodes[childNodeIdx], childNodeIdx, includedAttributes));
         }
         this.children = children;
     }
 
-    private static void setAttribute(Map<Attribute, Object> attribs, Attribute key, Object value) {
+    private boolean shouldIncludeAttribute(Attribute key) {
+        return includedAttributes.isEmpty() || includedAttributes.contains(key);
+    }
+
+    private void setAttribute(Map<Attribute, Object> attribs, Attribute key, Object value) {
         if (value != null) {
             attribs.put(key, value);
         }
     }
 
+    private Map<Attribute, Object> collectAttributes(AccessibilityNodeInfo node, int index) {
```

Comment:
Codacy found an issue: [The method 'collectAttributes(AccessibilityNodeInfo, int)' has an NPath complexity of 3670016, current threshold is 200](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=6310993)